primitives/core/src/traits.rs

```diff
@@ -69,12 +85,12 @@ pub trait BareCryptoStore: Send + Sync {
 	/// Find intersection between provided keys and supported keys
 	///
 	/// Provided a list of (CryptoTypeId,[u8]) pairs, this would return
-	/// a filtered list of public keys which are supported by the keystore.
-	fn supported_keys(&self, id: KeyTypeId, keys: Vec<CryptoTypePublicPair>) -> Result<Vec<CryptoTypePublicPair>, String>;
+	/// a filtered set of public keys which are supported by the keystore.
+	fn supported_keys(&self, id: KeyTypeId, keys: Vec<CryptoTypePublicPair>) -> Result<HashSet<CryptoTypePublicPair>, Error>;
```

Comment:
@tomusdrw did you really mean returning `HashSet` here? I understood that comment as checking against a `hashset` internally, but not returning one. I am also wondering, if we want to keep an ordered return value ... just considering that many implementations would just use the first one (which is not deterministically possible with a `HashSet`), an external signer could express the preference in keys through the order. When returning a `HashSet` it can't anymmore, it is pure luck (or it must return a `HashSet` of len 1 to express that preference). I am not sure how important that is and whether this should be implied or maybe even explicitly said in the docs ... 

I feel 6 out 10, that a Vec is the more flexible solution...

Reply:
Secondly just realised that for `sign_with` we only take a reference to `CryptoTypePublicPair`, while here we consume the entire type. Just thinking about the use case and am wondering, if we assumed we get the keys from on-chain storage, do we get them as references? If so, wouldn't the current API require an unneeded `clone`, while the following would not?

```suggestion
	fn supported_keys<'a>(&self, id: KeyTypeId, keys: Vec<&'a CryptoTypePublicPair>) -> Result<Vec<&'a CryptoTypePublicPair>, Error>;
```

:thinking: ?!?

Reply:
I didn't mean `HashSet` here, I suggested it might be useful for `fn keys` but at the same time we can just return `Vec`s everywhere and use `HashSet` internally everywhere (this would mean that `supported_keys` should not use `keys` and then convert from Vec to HashSet but rather copy the implementation and collect to a `HashSet`)

Reply:
I understand the concern is about the order of keys passed to this function is not preserved when using a HashSet. As far as i can see, even if we end up using the hashset internally, we would still lose the order unless we re-sort the keys:
```
let intersection = keys.intersection(...).....;
let result: Vec<CryptoTypePublicPair> = keys.into_iter().filter(|key| intersection.contains(key));
```

Does this sound good?

Reply:
I don't understand the `keys.intersection()` part. [`Intersection`](https://doc.rust-lang.org/std/collections/hash_set/struct.Intersection.html) internally is just a wrapper type, that then becomes an iterator (for our usecase). Considering that we aren't needing it, it is probably even more efficient to just do it bare:

```rust
fn supported_keys<'a>(&self, id: KeyTypeId, keys: Vec<&'a CryptoTypePublicPair>) -> Result<Vec<&'a CryptoTypePublicPair>, Error> {
	let supported_keys = self.iter_keys(id).collect::<HashSet<_>>() //internal function that returns an iterator
	keys.into_iter().filter(|key| supported_keys.contains(key).collect::<Vec<_>>()
}
```

Considering that `iter_keys()` is an internal function similarly to what `.keys()` does now, but returns an iterator rather than the collected `Vec` (for performance). `.keys()` would use the same `iter_keys()` internally to collect the `Vec`. Arguably this could even be part of the trait in `std` ... but I dunno if that is a good pattern (to have different trait interfaces during `std`/`no_std`)

Reply:
Why should order matter here?  We're never deterministic with signing anyways.

Reply:
Implemented @gnunicorn's comment about `HashSet`s but not including the `iter_keys` suggestion